src/common/hist_builder.h

```diff
@@ -0,0 +1,156 @@
+/*!
+ * Copyright 2017-2021 by Contributors
+ * \file hist_builder.h
+ */
+#ifndef XGBOOST_COMMON_HIST_BUILDER_H_
+#define XGBOOST_COMMON_HIST_BUILDER_H_
+
+#include <algorithm>
+#include <vector>
+#include "hist_util.h"
+#include "../data/gradient_index.h"
+
+#if defined(XGBOOST_MM_PREFETCH_PRESENT)
+  #include <xmmintrin.h>
+  #define PREFETCH_READ_T0(addr) _mm_prefetch(reinterpret_cast<const char*>(addr), _MM_HINT_T0)
+#elif defined(XGBOOST_BUILTIN_PREFETCH_PRESENT)
+  #define PREFETCH_READ_T0(addr) __builtin_prefetch(reinterpret_cast<const char*>(addr), 0, 3)
+#else  // no SW pre-fetching available; PREFETCH_READ_T0 is no-op
+  #define PREFETCH_READ_T0(addr) do {} while (0)
+#endif  // defined(XGBOOST_MM_PREFETCH_PRESENT)
+
+namespace xgboost {
+namespace common {
+
+struct Prefetch {
+ public:
+  static constexpr size_t kCacheLineSize = 64;
+  static constexpr size_t kPrefetchOffset = 10;
+
+ private:
+  static constexpr size_t kNoPrefetchSize =
+      kPrefetchOffset + kCacheLineSize /
+      sizeof(decltype(GHistIndexMatrix::row_ptr)::value_type);
+
+ public:
+  static size_t NoPrefetchSize(size_t rows) {
+    return std::min(rows, kNoPrefetchSize);
+  }
+
+  template <typename T>
+  static constexpr size_t GetPrefetchStep() {
+    return Prefetch::kCacheLineSize / sizeof(T);
+  }
+};
+
+template<typename FPType, bool do_prefetch,
+         typename BinIdxType, bool is_root,
+         bool any_missing>
+void BuildHistKernel(const std::vector<GradientPair>& gpair,
```

Comment:
Excellent that it's now in a header so we don't have to define those specializations!